src/core/lib/security/credentials/credentials.h

```diff
@@ -159,6 +161,22 @@ using CredentialsMetadataArray = std::vector<std::pair<Slice, Slice>>;
 
 /* --- grpc_call_credentials. --- */
 
+namespace grpc_core {
+
+// Abstract interface to obtain contextual data for some call credentials.
+class AuthMetadataContext {
```

Comment:
I wondered the same, but the guts of the two methods are almost identical - just returning different artifacts from the same computation, which seems to call for them being a little bit shared.

Reply:
We'd still be plumbing something through though, so I'm not convinced it actually simplifies things? And I think it sets the defaults wrong: now if there's a new thing that needs to be passed through we'll prefer to do it eagerly even if it's thrown away, whereas if we expose this interface the default is to expand the interface, preferring lazy computation so that only credentials that need the new thing do the work.